### PR TITLE
Keep scrollX in bounds

### LINKED ISSUE #1312: In safari on mac the waveform disappears when scrolling via the mouse wheel
### Wavesurfer.js version(s):
2.0.3

### Browser version(s):
Safari v11.0.1 on OSX v10.13.1

### Use behaviour needed to reproduce the issue:

Reproducible for me every time in Safari on macbook touchpad and mac mini magic trackpad.
Create a waveform with the param `scrollParent: true` and scroll along the x axis using the "mouse wheel".
If you scroll to the very end or back to the beginning the waveform will disappear and an error will log with further scrolling.
The error occurs in `fillRect` due to the index being out of bounds (there not being an entry canvas at -1 or >=canvases.length).
Safari has an elastic scroll when using the "mouse wheel" and allows for scrollX to be out of bounds.

To fix drawer.js line 223
```javascript
    getScrollX() {
        var maxScroll = this.params.scrollParent
            ? ~~(
                  this.wrapper.scrollWidth * this.params.pixelRatio -
                  this.getWidth()
              )
            : this.getWidth();

        return Math.min(
            maxScroll,
            Math.max(
                0,
                Math.round(this.wrapper.scrollLeft * this.params.pixelRatio)
            )
        );
    }
```

Will create a PR.

### PR DESCRIPTION
**Pull request by roboswank**
### Short description of changes:
Keep scrollX from going beyond the container or below 0.

### Notes:
The only disadvantage I see to this check happening here, is if someone is counting on special behavior when using elastic scroll beyond the container extents. In which case `drawBuffer` needs to be checking that `start`/`end` stay in bounds.

### Related Issues and other PRs:
fixes #1312

### CODE REVIEW

**Comment by thijstriemstra** [comment]
@roboswank can you also add this to the changelog and put a comment or two above your change explaining what it's doing?

**Comment by roboswank** [comment]
@thijstriemstra Will do.

**Comment by thijstriemstra** [comment]
Looks good to me, maybe @katspaugh can have the final say.

File: src/drawer.js [hunk 1]
```diff
@@ -220,7 +220,23 @@ export default class Drawer extends util.Observer {
...
     getScrollX() {
-        return Math.round(this.wrapper.scrollLeft * this.params.pixelRatio);
+        const pixelRatio = this.params.pixelRatio;
+        let x = Math.round(this.wrapper.scrollLeft * pixelRatio);
+
+        // In cases of elastic scroll (safari with mouse wheel) you can
+        // scroll beyond the limits of the container
+        // Calculate and floor the scrollable extent to make sure an out
+        // of bounds value is not returned
+        // https://github.com/katspaugh/wavesurfer.js/issues/1312
```
Comment (thijstriemstra) [inline src/drawer.js]: you dont have to include the full url, just a ticket reference should be fine.

Comment (roboswank) [reply src/drawer.js]: Gotcha, will update.

**Pull request merged**